contracts/NFTStorefrontV2.cdc

```diff
@@ -0,0 +1,705 @@
+
+            // Store the calculated sale price
+            self.salePrice = salePrice
+        }
+    }
+
+
+    // ListingPublic
+    // An interface providing a useful public interface to a Listing.
+    //
+    pub resource interface ListingPublic {
+        // borrowNFT
+        // This will assert in the same way as the NFT standard borrowNFT()
+        // if the NFT is absent, for example if it has been sold via another listing.
+        //
+        pub fun borrowNFT(): &NonFungibleToken.NFT
+
+        // purchase
+        // Purchase the listing, buying the token.
+        // This pays the beneficiaries and returns the token to the buyer.
+        //
+        pub fun purchase(
+            payment: @FungibleToken.Vault, 
+            commissionRecipient: Capability<&{FungibleToken.Receiver}>,
+            ): @NonFungibleToken.NFT
+
+        // getDetails
+        // Fetches the details of the listing.
+        pub fun getDetails(): ListingDetails
+
+    }
+
+
+    // Listing
+    // A resource that allows an NFT to be sold for an amount of a given FungibleToken,
+    // and for the proceeds of that sale to be split between several recipients.
+    // 
+    pub resource Listing: ListingPublic {
+        // The simple (non-Capability, non-complex) details of the sale
+        access(self) let details: ListingDetails
+
+        // A capability allowing this resource to withdraw the NFT with the given ID from its collection.
+        // This capability allows the resource to withdraw *any* NFT, so you should be careful when giving
+        // such a capability to a resource and always check its code to make sure it will use it in the
+        // way that it claims.
+        access(contract) let nftProviderCapability: Capability<&{NonFungibleToken.Provider, NonFungibleToken.CollectionPublic}>
+
+        // Optional list of marketplaces capabilities, Used for incentivising 
+        // them if purchase gets facilitated by them.
+        access(contract) let marketplacesCap: [Capability<&{FungibleToken.Receiver}>]?
+
+        // borrowNFT
+        // This will assert in the same way as the NFT standard borrowNFT()
+        // if the NFT is absent, for example if it has been sold via another listing.
+        //
+        pub fun borrowNFT(): &NonFungibleToken.NFT {
+            let ref = self.nftProviderCapability.borrow()!.borrowNFT(id: self.getDetails().nftID)
+            assert(ref.isInstance(self.getDetails().nftType), message: "NFT has wrong type")
+            assert(ref.id == self.getDetails().nftID, message: "NFT has wrong ID")
+            return ref as &NonFungibleToken.NFT
+        }
+
+        // getDetails
+        // Get the details of the current state of the Listing as a struct.
+        // This avoids having more public variables and getter methods for them, and plays
+        // nicely with scripts (which cannot return resources). 
+        //
+        pub fun getDetails(): ListingDetails {
+            return self.details
```

Comment:
What if an item is a ghost listing, this method will return that item as not purchased even if the underlying resource is gone.

if it is beeing sold at storefront that is handled but what if it is schwaped with somebody or sold using another contract?

Reply:
> What if an item is a ghost listing, this method will return that item as not purchased even if the underlying resource is gone.

Why would the underlying resource will gone? Can you please elaborate on this a little bit?

> if it is being sold at storefront that is handled but what if it is schwaped with somebody or sold using another contract?

If the NFT is not stored in the given provider capability anymore then the transaction would fail if someone tries to purchase. To clear that kind of orphan listing users should be aware of using the better UX tools as we discussed in the Nonce view issue thread.

Reply:
> Why would the underlying resource will gone? Can you please elaborate on this a little bit?

If i list and item for sale in storefront and then send it to another user. The item is gone from my storage so if I try to buy it it will panic. Right now when i get the listing information for that item it will show as present and everyting is fine. However it is not fine this item is a "ghost" it is not there.